docs/core/diagnostics/built-in-metrics-efcore.md

```diff
@@ -0,0 +1,69 @@
+---
+title: Entity Framework Core metrics
+description: Review the metrics available for Entity Framework Core
+ms.topic: reference
+ms.date: 10/06/2024
+---
+
+# Entity Framework Core metrics
+
+This article describes the metrics built-in for Entity Framework Core produced using the <xref:System.Diagnostics.Metrics?displayProperty=nameWithType> API.
+
+## `Microsoft.EntityFrameworkCore`
+
+The `Microsoft.EntityFrameworkCore` metrics report information about operations done by Entity Framework Core:
+
+- [`microsoft.entityframeworkcore.active_dbcontexts`](#metric-microsoftentityframeworkcoreactive_dbcontexts)
+- [`microsoft.entityframeworkcore.queries`](#metric-microsoftentityframeworkcorequeries)
+- [`microsoft.entityframeworkcore.savechanges`](#metric-microsoftentityframeworkcoresavechanges)
+- [`microsoft.entityframeworkcore.compiled_query_cache_hit_rate`](#metric-microsoftentityframeworkcorecompiled_query_cache_hit_rate)
+- [`microsoft.entityframeworkcore.execution_strategy_operation_failures`](#metric-microsoftentityframeworkcoreexecution_strategy_operation_failures)
+- [`microsoft.entityframeworkcore.optimistic_concurrency_failures`](#metric-microsoftentityframeworkcoreoptimistic_concurrency_failures)
+
+##### Metric: `microsoft.entityframeworkcore.active_dbcontexts`
+
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `microsoft.entityframeworkcore.active_dbcontexts` | ObservableUpDownCounter | `{dbcontext}` | Number of currently active `DbContext` instances. |
+
+Available starting in: Entity Framework Core 9.0.
+
+##### Metric: `microsoft.entityframeworkcore.queries`
+
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `microsoft.entityframeworkcore.queries` | ObservableCounter | `{query}` | Cumulative count of queries executed. |
+
+Available starting in: Entity Framework Core 9.0.
+
+##### Metric: `microsoft.entityframeworkcore.savechanges`
+
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `microsoft.entityframeworkcore.savechanges` | ObservableCounter | `{savechanges}` | Cumulative count of changes saved. |
+
+Available starting in: Entity Framework Core 9.0.
+
+##### Metric: `microsoft.entityframeworkcore.compiled_query_cache_hit_rate`
+
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `microsoft.entityframeworkcore.compiled_query_cache_hit_rate` | ObservableGauge | `%` | Hit rate - since last observation - for the compiled query cache. |
```

Comment:
I'm guessing "since last observation" means since the last time the ObservableGauge callback function was called? Not a doc issue, but that may be a functional/design issue for this metric. There is no guarantee that only one observer exists or what the relative timings are between observations. For example imagine the app developer sets up OpenTelemetry polling at 1 minute intervals, but then sometimes they also run dotnet-counters reporting at 1 second intervals. Your callback might get invoked:
```
t=0:01 (dotnet-counters)
t=0:02 (dotnet-counters)
...
t=0:59 (dotnet-counters)
t=1:00 (dotnet-counters)
t=1:00 (OpenTelemetry)
t=1:01 (dotnet-counters)
```
Those two calls at the 1 minute mark probably aren't going to be exactly the same time, maybe 10 microseconds happens to pass between them. OpenTelemetry is now reporting the cache hit rate for a period of time 10 microseconds long which might well be 0 hits of 0 attempts. 

You generally want to pick metrics that do not vary based on when the metric was last queried because that is an unpredictable time interval. You could report total hits and total attempts since the cache initialized and let users do a query to compute (hits_now-hits_last)/(total_now-total_last). Alternatively you could do the % hit rate over a time window that is not based on past observations. Hit rate since process start, hit rate in the last minute.

Reply:
Noted. I guess we have the same problem with "old" event counters as well. I added https://github.com/dotnet/efcore/issues/33959 to keep track of it.

Reply:
Yeah, I wouldn't worry about fixing the old event counters at this point, better just leave them as they are - but we should fix the new metric before releasing.